os/binfmt/libelf/libelf_addrenv.c

```diff
@@ -149,20 +149,47 @@ int elf_addrenv_alloc(FAR struct elf_loadinfo_s *loadinfo, size_t textsize, size
 	/* Allocate memory to hold the ELF image */
 
 #ifdef CONFIG_APP_BINARY_SEPARATION
-#ifdef CONFIG_DEBUG_MM_HEAPINFO
-	ARCH_GET_RET_ADDRESS
-	loadinfo->textalloc = (uintptr_t)mm_malloc(loadinfo->uheap, textsize + datasize, retaddr);
-#else
-	loadinfo->textalloc = (uintptr_t)mm_malloc(loadinfo->uheap, textsize + datasize);
-#endif
+        /* Allocate the RAM partition to load the app into */
+        loadinfo->binp->ramsize += loadinfo->textsize + loadinfo->rosize + loadinfo->datasize;
+
+        if (mm_allocate_ram_partition(&loadinfo->binp->ramstart, &loadinfo->binp->ramsize) < 0) {
+                berr("ERROR: Failed to allocate RAM partition\n");
+                return -ENOMEM;
+        }
+
+	if (loadinfo->textsize > loadinfo->rosize) {
+		unsigned long align_mask = loadinfo->textsize - 1;
+		loadinfo->textalloc = (loadinfo->binp->ramstart + align_mask) & ~align_mask;
+		align_mask = loadinfo->rosize - 1;
+		loadinfo->roalloc = (loadinfo->textalloc + loadinfo->textsize + align_mask) & ~align_mask;
+		loadinfo->dataalloc = loadinfo->roalloc + loadinfo->rosize;
+	} else {
+		unsigned long align_mask = loadinfo->rosize - 1;
+		loadinfo->roalloc = (loadinfo->binp->ramstart + align_mask) & ~align_mask;
+		align_mask = loadinfo->textsize - 1;
+		loadinfo->textalloc = (loadinfo->roalloc + loadinfo->rosize + align_mask) & ~align_mask;
+		loadinfo->dataalloc = loadinfo->textalloc + loadinfo->textsize;
+	}
+	loadinfo->binp->heapstart = loadinfo->dataalloc + datasize;
+
 #else
 	loadinfo->textalloc = (uintptr_t)kumm_malloc(textsize + datasize);
 #endif
 	if (!loadinfo->textalloc) {
+		berr("ERROR: Failed to allocate text section (size = %u)\n", textsize);
+		return -ENOMEM;
+	}
+
+	if (!loadinfo->dataalloc) {
```

Comment:
For `#ifndef CONFIG_APP_BINARY_SEPARATION` case, `loadinfo->dataalloc` and `loadinfo->roalloc` are not set/initialized well. Is it okay?

Or I think we should make `#ifdef CONFIG_APP_BINARY_SEPARATION` for `loadinfo->dataalloc` and `loadinfo->roalloc` cases.